In [1]:
# standard imports
import pandas as pd
import numpy as np
#warnings
import warnings
warnings.filterwarnings('ignore')

# Acquire data sets

In [2]:
# read in 2010 San Antonio census data
df_10 = pd.read_csv('ACS-1YR-2010.csv')
# read in 2011 San Antonio census data
df_11 = pd.read_csv('ACS-1YR-2011.csv')
# read in 2012 San Antonio census data
df_12 = pd.read_csv('ACS-1YR-2012.csv')
# read in 2013 San Antonio census data
df_13 = pd.read_csv('ACS-1YR-2013.csv')
# read in 2014 San Antonio census data
df_14 = pd.read_csv('ACS-1YR-2014.csv')
# read in 2015 San Antonio census data
df_15 = pd.read_csv('ACS-1YR-2015.csv')
# read in 2016 San Antonio census data
df_16 = pd.read_csv('ACS-1YR-2016.csv')
# read in 2017 San Antonio census data
df_17 = pd.read_csv('ACS-1YR-2017.csv')
# read in 2018 San Antonio census data
df_18 = pd.read_csv('ACS-1YR-2018.csv')
# read in 2019 San Antonio census data
df_19 = pd.read_csv('ACS-1YR-2019.csv')
# read in 2020 San Antonio census data
df_20 = pd.read_csv('ACS-5YR-2020.csv')
# read in 2021 San Antonio census data
df_21 = pd.read_csv('ACS-1YR-2021.csv')

# Prepare

To neatly combine data frames I want to see how many columns and rows each data frame contains. If they all contain the same number of columns and rows, I may be able to concat them together and then clean the single combined data set instead of multiple smaller data sets.

In [3]:
# find the number of columns in each data frame
# list of dataframes
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]
# for loop
for df in dataframes:
    # print the number of columns in each DataFrame
    num_columns = len(df.columns)
    print(f"Number of columns in the DataFrame: {num_columns}")
    # find the number of rows in each data frame
# list of dataframes
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]
# for loop
for df in dataframes:
    # print the number of rows in each DataFrame
    num_rows = len(df)
    print(f"Number of rows in the DataFrame: {num_rows}")

Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145


My data sets contain the same number of columns and rows. I will proceed with attempting to combine data frames. Because I anticipate the columns to be the same, I will concat the dataframes.

In [4]:
# concat df_10 and df_11
combined_df = pd.concat([df_10, df_11])
combined_df.shape

(290, 11)

When I attempt to concat df_10 and df_11 the data frames combine to 290 rows and 11 columns instead of the expected 145 rows and 9 columns. From a quick but thorough look at the csv files, all the columns and rows should match. It is possible they are not being read in the same? I will clean each data set individualy then combine into a single dataframe.

In [5]:
df_10.head()

Label (Grouping) Texas!!Estimate  \
0                 EMPLOYMENT STATUS             NaN   
1      Population 16 years and over      19,123,208   
2                    In labor force      12,465,332   
3              Civilian labor force      12,363,612   
4                          Employed      11,271,851   

  Texas!!Estimate Margin of Error Texas!!Percent  \
0                             NaN            NaN   
1                         ±10,217     19,123,208   
2                         ±33,691         65.20%   
3                         ±33,592         64.70%   
4                         ±36,125         58.90%   

  Texas!!Percent Margin of Error San Antonio city, Texas!!Estimate  \
0                            NaN                               NaN   
1                            (X)                         1,019,273   
2                           ±0.2                           660,424   
3                           ±0.2                           648,636   
4                           ±0.2                           589,877   

  San Antonio city, Texas!!Estimate Margin of Error  \
0                                               NaN   
1                                            ±5,093   
2                                            ±7,817   
3                                            ±8,332   
4                                            ±8,640   

  San Antonio city, Texas!!Percent  \
0                              NaN   
1                        1,019,273   
2                           64.80%   
3                           63.60%   
4                           57.90%   

  San Antonio city, Texas!!Percent Margin of Error  
0                                              NaN  
1                                              (X)  
2                                             ±0.7  
3                                             ±0.7  
4                                             ±0.8

Several changes need to be made to column names for readablity and simplicity. I will be formating column names to be lowercase, snakecase, abreviated, and I will remove any unecessary charachters.

In [6]:
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]
# for loop
for df in dataframes:
    # format column names to be lowercase
    df.columns = df.columns.str.lower()
    # format column names to be snakecase
    df.columns = df.columns.str.replace(' ', '_')
    # replace !! with snakecase
    df.columns = df.columns.str.replace('!!', '_')
    # remove commas
    df.columns = df.columns.str.replace(',', '')
    # shorten texas to tx
    df.columns = df.columns.str.replace('texas', 'tx')
    # shorten san antonio to sa
    df.columns = df.columns.str.replace('san_antonio', 'sa')
    # shorten margin of error to moe
    df.columns = df.columns.str.replace('margin_of_error', 'moe')
    # shorten estimate to est
    df.columns = df.columns.str.replace('estimate', 'est')
    # remove (
    df.columns = df.columns.str.replace('(', '')
    # remove )
    df.columns = df.columns.str.replace(')', '')

### original data frames
    df_10     df_16
    df_11     df_17
    df_12     df_18
    df_13     df_19
    df_14     df_20
    df_15     df_21
    
### sub category data frames
##### df_10
    df10_employment_status
    df10_commuting_to_work
    df10_occupation
    df10_industry
    df10_worker_class
    df10_income_and_benefits
    df10_health_insurance
    df10_poverty_line
##### df_11
    df11_employment_status
    df11_commuting_to_work
    df11_occupation
    df11_industry
    df11_worker_class
    df11_income_and_benefits
    df11_health_insurance
    df11_poverty_line
##### df_12
    df12_employment_status
    df12_commuting_to_work
    df12_occupation
    df12_industry
    df12_worker_class
    df12_income_and_benefits
    df12_health_insurance
    df12_poverty_line
##### df_13
    df13_employment_status
    df13_commuting_to_work
    df13_occupation
    df13_industry
    df13_worker_class
    df13_income_and_benefits
    df13_health_insurance
    df13_poverty_line
##### df_14
    df14_employment_status
    df14_commuting_to_work
    df14_occupation
    df14_industry
    df14_worker_class
    df14_income_and_benefits
    df14_health_insurance
    df14_poverty_line
##### df_15
    df15_employment_status
    df15_commuting_to_work
    df15_occupation
    df15_industry
    df15_worker_class
    df15_income_and_benefits
    df15_health_insurance
    df15_poverty_line
##### df_16
    df16_employment_status
    df16_commuting_to_work
    df16_occupation
    df16_industry
    df16_worker_class
    df16_income_and_benefits
    df16_health_insurance
    df16_poverty_line
##### df_17
    df17_employment_status
    df17_commuting_to_work
    df17_occupation
    df17_industry
    df17_worker_class
    df17_income_and_benefits
    df17_health_insurance
    df17_poverty_line
##### df_18
    df18_employment_status
    df18_commuting_to_work
    df18_occupation
    df18_industry
    df18_worker_class
    df18_income_and_benefits
    df18_health_insurance
    df18_poverty_line
##### df_19
    df19_employment_status
    df19_commuting_to_work
    df19_occupation
    df19_industry
    df19_worker_class
    df19_income_and_benefits
    df19_health_insurance
    df19_poverty_line
##### df_20
    df20_employment_status
    df20_commuting_to_work
    df20_occupation
    df20_industry
    df20_worker_class
    df20_income_and_benefits
    df20_health_insurance
    df20_poverty_line
##### df_21
    df21_mployment_status
    df21_commuting_to_work
    df21_occupation
    df21_industry
    df21_worker_class
    df21_income_and_benefits
    df21_health_insurance
    df21_poverty_line

In [7]:
# Assuming you have a list of DataFrames called df_list
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]

# Create a new list to store the updated DataFrames
updated_df_list = []

# Iterate over the DataFrames
for i, df in enumerate(dataframes):
    # Create a new list to store the updated column names
    new_column_names = []
    
    # Calculate the suffix value
    suffix = 10 + i

    # Iterate over the existing column names
    for column in df.columns:
        new_column_names.append(column + '_' + str(suffix))

    # Assign the updated column names to the DataFrame
    df.columns = new_column_names

    # Append the updated DataFrame to the list
    updated_df_list.append(df)

In [8]:
df_10.head()

label_grouping_10   tx_est_10 tx_est_moe_10 tx_percent_10  \
0                 EMPLOYMENT STATUS         NaN           NaN           NaN   
1      Population 16 years and over  19,123,208       ±10,217    19,123,208   
2                    In labor force  12,465,332       ±33,691        65.20%   
3              Civilian labor force  12,363,612       ±33,592        64.70%   
4                          Employed  11,271,851       ±36,125        58.90%   

  tx_percent_moe_10 sa_city_tx_est_10 sa_city_tx_est_moe_10  \
0               NaN               NaN                   NaN   
1               (X)         1,019,273                ±5,093   
2              ±0.2           660,424                ±7,817   
3              ±0.2           648,636                ±8,332   
4              ±0.2           589,877                ±8,640   

  sa_city_tx_percent_10 sa_city_tx_percent_moe_10  
0                   NaN                       NaN  
1             1,019,273                       (X)  
2                64.80%                      ±0.7  
3                63.60%                      ±0.7  
4                57.90%                      ±0.8

I want column: label_grouping to become my x-axis, so It will be easier to combine data frames, but first I need to clean the column values that will be my y-axis labels

In [9]:
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]

# Iterate over the dataframes
for i, df in enumerate(dataframes):
    # Convert row values to lowercase
    df.iloc[:, 0] = df.iloc[:, 0].str.lower()
    # Convert row values to snake_case
    df.iloc[:, 0] = df.iloc[:, 0].str.replace(' ', '_')
    # Remove -
    df.iloc[:, 0] = df.iloc[:, 0].str.replace('-', '')
    # Remove (
    df.iloc[:, 0] = df.iloc[:, 0].str.replace('(', '')
    # Remove )
    df.iloc[:, 0] = df.iloc[:, 0].str.replace(')', '')
    # Remove commas
    df.iloc[:, 0] = df.iloc[:, 0].str.replace(',', '')
    # Remove :
    df.iloc[:, 0] = df.iloc[:, 0].str.replace(':', '')
    # Remove /
    df.iloc[:, 0] = df.iloc[:, 0].str.replace('/', '')

Because this is a large data set it will be easier to view the data if I create mini data frames of the data by isolating the original datframe by the subcatagories: employment status, commuting to work, occupation, industry, class of worker, income and benefits, health insurance, poverty line

In [10]:
# create dataframe for subcategory employment status
df10_employment_status = df_10[1:18]
# create dataframe for subcategory commuting to work
df10_commuting_to_work = df_10[19:27]
# create dataframe for subcategory occupation
df10_occupation = df_10[28:34]
# create dataframe for subcategory industry
df10_industry = df_10[35:49]
# create dataframe for subcategory class of worker
df10_worker_class = df_10[50:55]
# create dataframe for subcategory income and benefits
df10_income_and_benefits = df_10[56:100]
# create dataframe for subcategory health insurance
df10_health_insurance = df_10[101:125]
# create dataframe for subcategory poverty line
df10_poverty_line = df_10[126:145]

In [11]:
# create dataframe for subcategory employment status
df11_employment_status = df_11[1:18]
# create dataframe for subcategory commuting to work
df11_commuting_to_work = df_11[19:27]
# create dataframe for subcategory occupation
df11_occupation = df_11[28:34]
# create dataframe for subcategory industry
df11_industry = df_11[35:49]
# create dataframe for subcategory class of worker
df11_worker_class = df_11[50:55]
# create dataframe for subcategory income and benefits
df11_income_and_benefits = df_11[56:100]
# create dataframe for subcategory health insurance
df11_health_insurance = df_11[101:125]
# create dataframe for subcategory poverty line
df11_poverty_line = df_11[126:145]

In [12]:
# create dataframe for subcategory employment status
df12_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df12_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df12_occupation = df_12[28:34]
# create dataframe for subcategory industry
df12_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df12_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df12_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df12_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df12_poverty_line = df_12[126:145]

In [13]:
# create dataframe for subcategory employment status
df13_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df13_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df13_occupation = df_12[28:34]
# create dataframe for subcategory industry
df13_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df13_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df13_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df13_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df13_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df14_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df14_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df14_occupation = df_12[28:34]
# create dataframe for subcategory industry
df14_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df14_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df14_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df14_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df14_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df15_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df15_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df15_occupation = df_12[28:34]
# create dataframe for subcategory industry
df15_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df15_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df15_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df15_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df15_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df16_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df16_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df16_occupation = df_12[28:34]
# create dataframe for subcategory industry
df16_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df16_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df16_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df16_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df16_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df17_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df17_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df17_occupation = df_12[28:34]
# create dataframe for subcategory industry
df17_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df17_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df17_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df17_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df17_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df18_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df18_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df18_occupation = df_12[28:34]
# create dataframe for subcategory industry
df18_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df18_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df18_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df18_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df18_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df19_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df19_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df19_occupation = df_12[28:34]
# create dataframe for subcategory industry
df19_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df19_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df19_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df19_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df19_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df20_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df20_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df20_occupation = df_12[28:34]
# create dataframe for subcategory industry
df20_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df20_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df20_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df20_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df20_poverty_line = df_12[126:145]

In [ ]:
# create dataframe for subcategory employment status
df21_employment_status = df_12[1:18]
# create dataframe for subcategory commuting to work
df21_commuting_to_work = df_12[19:27]
# create dataframe for subcategory occupation
df21_occupation = df_12[28:34]
# create dataframe for subcategory industry
df21_industry = df_12[35:49]
# create dataframe for subcategory class of worker
df21_worker_class = df_12[50:55]
# create dataframe for subcategory income and benefits
df21_income_and_benefits = df_12[56:100]
# create dataframe for subcategory health insurance
df21_health_insurance = df_12[101:125]
# create dataframe for subcategory poverty line
df21_poverty_line = df_12[126:145]

Subcategory data frames are complete. The changes in the below cell still need to be made.

null and (X) values need to be delt with
values need commas, +-, and % removed.

maybe drop moe columns

In [7]:
#The data sets have a x-axis label but the label was read in as a column. The column: Label (Grouping), has been set as the index.
# set the index of each data frame to be column: 'Label (Grouping)'
#df_10 = df_10.set_index('Label (Grouping)')
#df_11 = df_11.set_index('Label (Grouping)')
#df_12 = df_12.set_index('Label (Grouping)')
#df_13 = df_13.set_index('Label (Grouping)')
#df_14 = df_14.set_index('Label (Grouping)')
#df_15 = df_15.set_index('Label (Grouping)')
#df_16 = df_16.set_index('Label (Grouping)')
#df_17 = df_17.set_index('Label (Grouping)')
#df_18 = df_18.set_index('Label (Grouping)')
#df_19 = df_19.set_index('Label (Grouping)')
#df_20 = df_20.set_index('Label (Grouping)')
#df_21 = df_21.set_index('Label (Grouping)')


#There are several labels used to label subgroups in the dataframe but do not contain information its self. I will drop these below.
# drop EMPLOYMENT STATUS
#df_10.drop(labels='EMPLOYMENT STATUS', inplace=True)
# drop COMMUTING TO WORK
#df_10.drop(labels='COMMUTING TO WORK', inplace=True)
# drop OCCUPATION
#df_10.drop(labels='OCCUPATION', inplace=True)
# drop INDUSTRY
#df_10.drop(labels='INDUSTRY', inplace=True)
# drop CLASS OF WORKER
#df_10.drop(labels='CLASS OF WORKER', inplace=True)
# drop INCOME AND BENEFITS (IN 2010 INFLATION-ADJUSTED DOLLARS)
#df_10.drop(labels='INCOME AND BENEFITS (IN 2010 INFLATION-ADJUSTED DOLLARS)', inplace=True)
# drop HEALTH INSURANCE COVERAGE
#df_10.drop(labels='HEALTH INSURANCE COVERAGE', inplace=True)
# drop PERCENTAGE OF FAMILIES AND PEOPLE WHOSE INCOME IN THE PAST 12 MONTHS IS BELOW THE POVERTY LEVEL
#df_10.drop(labels='PERCENTAGE OF FAMILIES AND PEOPLE WHOSE INCOME IN THE PAST 12 MONTHS IS BELOW THE POVERTY LEVEL', inplace=True)